### ISSUE TITLE
Knowledge Gap: Cloud Role Name

### ISSUE THREAD

**Issue by MS-TimothyMothra** [Microsoft/ApplicationInsights-Home#326]
I've seen an increase in customer issues over the use and configuration of Cloud Role Name.
This is a huge knowledge gap that needs to be addressed.
Specifically what values are being captured by default by our SDKs and how those SDKs behave differently **on-prem, on Azure, and On-By-Default environments**.

Collecting information here:

- Application Map has excellent examples here: https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-map#set-cloudrolename

- Dot Net Web SDK will attempt to autopopulate this when hosted in Azure
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/96cb36e0b360d7e99af9f8c9b36f7abe1c8e344e/Src/WindowsServer/WindowsServer.Shared/Implementation/AzureServiceRuntimeAssemblyLoader.cs#L24-L57

**Comment by MS-TimothyMothra** [comment]
We keep getting requests to customize the cloud_RoleName or appName without changing code.
Specifically, with our code-less attach scenarios.

One recommendation is to introduce an environment variable that can be set on a machine to allow customers set this.

Pro:
- codeless solution for codeless attach.

Con:
- if this is an environment variable, this may conflict with customers who run multiple apps on one machine.

**Comment by cijothomas** [comment]
@MS-TimothyMothra instead of ENV, can we get RoleName from Web.Config (apsettings.json in .net core) ? This will avoid unintended side effects..

**Comment by paulomorgado** [comment]
I have a [draft PR](https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/1158) to pull **cloud_RoleName** from the application pool name.

But I would favor a generic implementation that pulls properties out of ApplicationInsigths.config with environment variable expansion.

This could be an easy way to populate whatever the user needs:

```xml
<ApplicationInsights xmlns="http://schemas.microsoft.com/ApplicationInsights/2013/Settings">
  <TelemetryInitializers>
    <Add Type="ConfigurableInitializer">
      <RoleName>%APP_POOL_ID%</RoleName>
      <RoleInstance>%COMPUTERNAME%</RoleInstance>
    </Add>
  </TelemetryInitializers>
</ApplicationInsights>
```

**Comment by MS-TimothyMothra** [comment]
@cijothomas I'm hearing people specifically ask for an Environment Variable, but I definitely prefer your suggestion.

I would be in favor of a ranking of options for example: EnvVar < Web.Config < ApplicationInsights.config

I like @paulomorgado 's proposal.

**Comment by cijothomas** [comment]
@MS-TimothyMothra can customer modify applicationinsights.config file in codeless attach?

**Comment by paulomorgado** [comment]
I, actually, like both.

**Comment by paulomorgado** [comment]
@cijothomas,

> @MS-TimothyMothra can customer modify applicationinsights.config file in codeless attach?

If you have access to the system, you sure have access to ApplicationInsigths.config and that's one of the use cases.

**Comment by paulomorgado** [comment]
Draft PR that pulls **RoleName** and **RoleInstance** from the configuration of the telemetry initializer: https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/1161

**Comment by cijothomas** [comment]
> @cijothomas,
> 
> > @MS-TimothyMothra can customer modify applicationinsights.config file in codeless attach?
> 
> If you have access to the system, you sure have access to ApplicationInsigths.config and that's one of the use cases.

I am not sure if that's the case. We don't allow customers to edit ai.config if using Web App extensions. Not sure with Status Monitor - but as you said nothing would prevent one from modifying it if access.

**Comment by MS-TimothyMothra** [comment]
For the new Status Monitor, the ai.config is an embedded resource so it won't be available to end users.
The plan is to make certain options available through StatusMonitor as needed.

**Comment by cijothomas** [comment]
Since ai.config is NOT editable by user - I don't see how @paulomorgado submitted PR for new TelemetryInitializer about IIS  would help. Because user's won't be able to take advantage of it without accessing ai.config.

**Comment by MS-TimothyMothra** [comment]
I'm not opposed to Paulo's PR in general because I've seen this request for a lot of customers.

@cijothomas Where would the appropriate place for an EnvironmentVariable reader be?
I think we could add that in the 2.10 milestone.

**Comment by cijothomas** [comment]
How does users add Paulo's new TelemetryInitiazlier to the Ai.Config if we dont allow users to edit ai.Config?

**Comment by MS-TimothyMothra** [comment]
This issue isn't exclusive to codeless-attach.

But we should definitely prioritize the EnvironmentVariable solution if we want to get it into 2.10.

**Comment by cijothomas** [comment]
If its not for code-less scenarios, then user's can easily write TelemetryInitializers themselves , and we can put Paulo's code in examples repo for users.

**Comment by MS-TimothyMothra** [comment]
That's fine. Can we discuss where's the correct place to implement the EnvironmentVariable?

**Comment by paulomorgado** [comment]
My initial use case was an organization that manages the infrastructure for another organization that uses a third party developed application.

This all runs on VMs and there's definitely access to ApplicationInsights.config but not access to the source code of the application or anyone capable of writing the code, even from a sample.

This is a use case that covers both virtual or physical machines on-prem or on any cloud provider.

**Comment by cijothomas** [comment]
@paulomorgado If your use case is with StatusMonitor customers, this change will not help you because
1. if you are on OLD version of SM, it can only update to 2.3 of SDK which wont contain this proposed change.
2. If you are shifting to new SM, then as Timothy said, you wont get access to Ai.COnfig.

Let me know if my understanding is correct.

**Comment by paulomorgado** [comment]
Yes you do, @cijothomas. It's me that wasn't understanding that this is an all new SM.

But this is still a valid use case. What is your suggestion for this use case?

**Comment by cijothomas** [comment]
It looks like the issue is specific to SM customers. @MS-TimothyMothra can we solve this with SM itself? Like allow users to set role name for each app, just like they set ikey?

**Comment by paulomorgado** [comment]
That would be great!

**Comment by MS-TimothyMothra** [comment]
We're happy to take feature requests for SM.
But first, we need to support this at the SDK level first before I can support it in SM.

**Comment by SergeyKanzhelev** [comment]
Env vars would be better for this. See https://github.com/Microsoft/ApplicationInsights-dotnet/issues/630 and https://github.com/census-instrumentation/opencensus-specs/blob/b977bcf382470647fefd0ae2a699561eb3ccc929/resource/Resource.md#from-environment-variables

**Comment by cijothomas** [comment]
@MS-TimothyMothra I guess we can create a TelemetryInitializer which sets RoleName,RoleInstance from pre-defined ENV variables. And add this to the default list of initializers, before any other Azure specific initializers. (the 1st one always wins)

Do you want to post an issue with the description in dotnet repo to gain any additional feedbacks?

**Comment by paulomorgado** [comment]
https://github.com/microsoft/ApplicationInsights-dotnet-server/pull/1161

**Comment by ketaanhshah** [comment]
Hi All,
anyone know how to set RoleName (Cloud RoleName) for Azure API Management (APIM) instance created? We dont want to create a separate Application Insights for the API Management. I would like to know how to set the RoleName for APIM as we can do for custom application thru telemetry initializer how we do for APIM>

**Comment by cijothomas** [comment]
@ketaanhshah Dont think api management allows this option as of now.

**Comment by cijothomas** [comment]
@paulomorgado We are having a discussion internally about this. will post back summary. Your PR will be reviewed along with that.

**Comment by ketaanhshah** [comment]
@cijothomas Hello, do we have any inputs on APIM Cloud RoleName? Can we use Custom Telemetry for setting cloud roleName and also currently we have a requirement where the APIM is logging PII data and that needs to be masked. How can we mask that Telemetry coming thry APIM into Azure, for other applications we are able to use CustomTelemetry class and control it, can we do the same for APIM? Any help?

**Comment by ketaanhshah** [comment]
@Dmitry-Matveev Hello Dmity, we have a PII masking requirement and I happen to come across your post here (https://social.msdn.microsoft.com/Forums/azure/en-US/0b38fd1e-8aa9-45f7-91a7-fd0631ef8bba/dealing-with-pii-or-sensitive-data-captured-by-application-insights?forum=ApplicationInsights) My question is how we do this for API Management (APIM)? As I am not sure how to associate the Custom Telemetry with API Management. As right now its all configured thru Azure Portal and no custom telemetry to it, our backend services (API) do use Custom Telemetry but in Azure portal the PII data is marked as coming from APIM and not the APIs itself. Any help?

**Comment by cijothomas** [comment]
@ketaanhshah as we mentioned earlier, APIM does not support this (support adding telemetry initializer).

**Comment by ketaanhshah** [comment]
@cijothomas Thank you for that then how do we MANAGE the PII data thats being logged in App Insights and is coming from APIM? any other solution for masking such PII data @Dmitry-Matveev

**Comment by cijothomas** [comment]
This must be asked in APIM forum/docs page to get help.

**Comment by ketaanhshah** [comment]
@cijothomas but this Application Insights issue not an APIM issue, APIM is sending data and like @Dmitry-Matveev pointed out in his thread a custom Telemetry implementation is required for it but how do we configure it in APIM. I will post the same question in APIM too but thought App Insights experts may have some insights on it.

**Comment by cijothomas** [comment]
For application insights, the answer is always Telemetryprocessor/TelemetryInitializer to filter out PII data. APIM is not allowing additional of initializer/processor, and hence this question is best answered in  APIM github, not in application insights.

**Comment by ketaanhshah** [comment]
@cijothomas Thank you, have posted to APIM forum and I get some solution to it as exposing PII data is really bad and dont know how else this can be configured as documentation lacks it and so does the response from APIM team so far.
